### PR TITLE
Added a ChasingElement class using the Observer pattern; updated rela…

### PR DESCRIPTION
**Pull request by gwengou**
…ted classes accordingly. Added multiple utility methods to Point2D class. Embedded a ChasingElement in GraphicsMain for testing purposes.

### CODE REVIEW

File: src/main/java/elements/MovableElement.java
Comment (AditMeh) [inline src/main/java/elements/MovableElement.java:70]: just use this.velocity instead of the getter

File: src/main/java/game/Game.java
Comment (AditMeh) [inline src/main/java/game/Game.java:68]: Violates clean architecture and single responsibility.

-  Don't bring boards and object managers/objects (entities) into game (controller)
-  Game is responsible for connecting to player and board. Not for updating the playerstate.

This code actually belongs in objectManager, all this function should call is line 79 below.

File: src/main/java/game/ObjectManager.java
Comment (AditMeh) [inline src/main/java/game/ObjectManager.java:39]: See my comment on game.java. This doesn't really need to exist.

File: src/main/java/elements/ChasingElement.java
Comment (AditMeh) [inline src/main/java/elements/ChasingElement.java:63]: Shouldn't it be super.<method>, instead of this.method? Or am I just being dumb here.

File: src/main/java/graphics/GraphicsMain.java
Comment (AditMeh) [inline src/main/java/graphics/GraphicsMain.java:114]: Again, I don't think it's clean to continuously import from bottom to top. We can just do game.addChasingElement(), which will then call all the required methods.

File: src/main/java/utils/EnumsForSprites.java
Comment (AditMeh) [inline src/main/java/utils/EnumsForSprites.java:3]: Might as well make chaser capitalized like CHASER, to follow suit

**Review by AditMeh — CHANGES_REQUESTED** [review]
Just some minor changes to be made. This may be a bit of an ask, but is it fine if you implement a reset() method (as outlined [here](https://github.com/CSC207-UofT/course-project-1190/blob/make_play_mode/src/main/java/elements/MovableElement.java)). I essentially want all the elements to be able to be reset to some "initial" state (you should store this). So in this case, maybe you need to reset the velocity and position of the movable object to some stored defaults.

File: src/main/java/elements/MovableElement.java
Comment (gwengou) [reply src/main/java/elements/MovableElement.java:70]: For the ChasingElement to follow the Player, its velocity needs to change and we need to get every update instantly. This.velocity does not reflect the change....maybe because it was memorized at instantiation? Not sure, but it did not work.

File: src/main/java/game/Game.java
Comment (gwengou) [reply src/main/java/game/Game.java:68]: I agree this is problematic. My problem was that the Player needs to add every ChasingElement as its Observer, but ObjectManager is agnostic about Player. The only place where Player and board objects interact is in the Game....What do you suggest?

File: src/main/java/elements/ChasingElement.java
Comment (gwengou) [reply src/main/java/elements/ChasingElement.java:63]: You mean this.setVelocity()? Sure but what's the difference....ChasingElement inherits setVelocity() from MovableElement, so its this.setVelocity() should be equivalent to super.setVelocity(), right?

File: src/main/java/elements/MovableElement.java
Comment (AditMeh) [reply src/main/java/elements/MovableElement.java:70]: I’ll take a look on my end then.

File: src/main/java/graphics/GraphicsMain.java
Comment (gwengou) [reply src/main/java/graphics/GraphicsMain.java:114]: Yes. I have said before it is not good Java style to chain the calls, but this embedded test section has all the GraphicsMain-level calls in this style....Is it correct to say that the GameMaker will call Game methods to plant the elements? If so, I'll try to wrap them all into Game-level methods.

File: src/main/java/utils/EnumsForSprites.java
Comment (gwengou) [reply src/main/java/utils/EnumsForSprites.java:3]: Yes, good call. Will fix.

File: src/main/java/game/Game.java
Comment (AditMeh) [reply src/main/java/game/Game.java:68]: Hmm I see, this is tough. I think the observer pattern itself breaks clean architecture since it needs to interact with  game. Let’s ask the TA.

File: src/main/java/elements/ChasingElement.java
Comment (AditMeh) [reply src/main/java/elements/ChasingElement.java:63]: I see. So let me know if this is correct: we cannot access instance variables inherited from the parent using this.<variable>. But we can use methods from the parent using this.<method>

Comment (gwengou) [reply src/main/java/elements/ChasingElement.java:63]: If the parent's instance variable has private access, children can't access it directly. Parent methods usually have public access, so children can access them using this.

File: src/main/java/graphics/GraphicsMain.java
Comment (AditMeh) [reply src/main/java/graphics/GraphicsMain.java:114]: Yea we’re going to have to change the calls structured in this style to calling game methods. So might as well start now. It’d be nice if you could also do it for the other methods as well.

File: src/main/java/elements/ChasingElement.java
Comment (AditMeh) [reply src/main/java/elements/ChasingElement.java:63]: I see, that makes sense. Thanks!

File: src/main/java/graphics/GraphicsMain.java
Comment (gwengou) [reply src/main/java/graphics/GraphicsMain.java:114]: Will do. I'm thinking it may be useful to create a separate GameSeeder class that handles all the element-planting.

**Comment by gwengou** [comment]
> Just some minor changes to be made. This may be a bit of an ask, but is it fine if you implement a reset() method (as outlined [here](https://github.com/CSC207-UofT/course-project-1190/blob/make_play_mode/src/main/java/elements/MovableElement.java)). I essentially want all the elements to be able to be reset to some "initial" state (you should store this). So in this case, maybe you need to reset the velocity and position of the movable object to some stored defaults.

Will have to wait until your branch is merged into main...ChasingElement will inherit that location reset() from MovableElement, with an additional velocity reset, but doing so on my end will unseemly merge conflicts.

**Comment by AditMeh** [comment]
I tested your code out locally on my computer. The chasing elements worked pretty well, they made my heart rate rise lol. 

One issue, if I move the player in a certain way, the chasing elements run off the board and don't come back. 

https://gyazo.com/38a0782e37ee24ea76fdd50e577cd2d6

**Comment by gwengou** [comment]
The velocity of ChasingElements change only if the Player's position changes. When you hit the left wall, Player position is  frozen, and the monsters are going to keep moving in the direction they have been. In a sense, you can see it as the monsters becoming "blind" to a non-moving Player; one strategy for Player to dodge chasing monsters is to stay put somewhere until the monsters pass/travel far enough...

**Comment by kakaomonk** [comment]
I went over your code. Probably I should change some logic of Pushable because of gameSeeder.java, but it wouldn't take too long time. Just let me know after you merge your branch into the main.

**Pull request merged**
